File: cpp/2022/day8.cpp
#include <bits/stdc++.h>
using namespace std;

vector<vector<int>> hg;
vector<vector<bool>> vi;
int n;

void look1(int x, int y, int dx, int dy) {
    int m = -1;
    for (int i = x, j = y; i >= 0 && j >= 0 && i < n && j < n; i += dx, j += dy) {
        if (m >= hg[i][j]) continue;
        vi[i][j] = 1;
        m = hg[i][j];
    }
}

int look2(int x, int y, int dx, int dy) {
    int h = hg[x][y], c = 0;
    for (int i = x + dx, j = y + dy; i >= 0 && j >= 0 && i < n && j < n; i += dx, j += dy) {
        c++;
        if (h <= hg[i][j]) break;
    }
    return c;
}

int main() {
    ifstream f("input.txt");
    string str;
    for (int i = 0; getline(f, str); i++) {
        hg.resize(str.size());
        hg[i].resize(str.size());
        for (int j = 0; j < str.size(); j++) hg[i][j] = str[j] - '0';
    }
    n = hg.size();
    vi.resize(n, vector<bool>(n));
    for (int i = 0; i < n; i++) {
        look1(i, 0, 0, 1);
        look1(i, n - 1, 0, -1);
        look1(0, i, 1, 0);
        look1(n - 1, i, -1, 0);
    }
    int s = 0;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            s += vi[i][j];
        }
    }
    int ma = 0;
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            ma = max(ma, look2(i, j, 0, 1) * look2(i, j, 1, 0) * look2(i, j, 0, -1) * look2(i, j, -1, 0));
        }
    }
    cout << s << endl << ma;
}

File: haskell/2022/day8.hs
import Data.List (transpose, reverse, inits)

zipSelf op x = zipWith op x (tail x)

islvis :: [[Int]] -> [[Bool]]
islvis = map ((True:) . zipSelf (<) . scanl1 max)

lviews :: [[Int]] -> [[Int]]
lviews = map (map viewFromLast . tail . inits)

countUntilOrLen :: Int -> [Int] -> Int
countUntilOrLen _ [] = 0 
countUntilOrLen n (x:xs)
    | x >= n = 1
    | otherwise = 1 + countUntilOrLen n xs

viewFromLast :: [Int] -> Int
viewFromLast x = countUntilOrLen t rr 
        where (t:rr) = reverse x

main = do
    brd <- map (map $ read . (:[])) . lines <$> readFile "input.txt" :: IO [[Int]]
    let a1 = islvis brd
    let b1 = transpose $ islvis $ transpose $ brd
    let c1 = map reverse $ islvis $ map reverse $ brd
    let d1 = reverse . transpose $ islvis $ transpose . reverse $ brd
    print $ sum . map (length . filter id) $ foldr1 (zipWith $ zipWith (||)) [a1, b1, c1, d1]
    let a2 = lviews brd
    let b2 = transpose $ lviews $ transpose $ brd
    let c2 = map reverse $ lviews $ map reverse $ brd
    let d2 = reverse . transpose $ lviews $ transpose . reverse $ brd
    print $ maximum . map maximum $ foldr1 (zipWith $ zipWith (*)) [a2, b2, c2, d2]